### DISCUSSION TITLE
Issue with Enabling Shaderc, D3D11, and Vulkan Features in MPV Cross Compilation on Ubuntu

### DISCUSSION THREAD

**Discussion by Windev2024** [mpv-player/mpv#14389] [Q&A]
**System Details:**

    Host OS: Ubuntu 22.04
    Target OS: Windows (cross-compiling)
    Cross-Compiler: MinGW (x86_64-w64-mingw32)
    
 
   **Description:**
I am following the steps outlined in the MPV documentation for cross-compiling on Windows: [MPV Compilation Guide](https://github.com/mpv-player/mpv/blob/master/DOCS/compile-windows.md). I successfully built 32-bit and 64-bit versions of MPV, and the standalone MPV is able to play. However, I need to add full-fledged features, and I am encountering issues enabling certain features.

**Meson Setup Command:**
`meson setup --wrap-mode=forcefallback -Ddefault_library=static -Dprefer_static=true \
    -Dc_link_args='-static' -Dcpp_link_args='-static' \
    -Dffmpeg:default_library=static -Dffmpeg:tests=disabled -Dffmpeg:programs=disabled -Dffmpeg:sdl2=disabled -Dffmpeg:vulkan=auto \
    -Dlcms2:fastfloat=true -Dlcms2:jpeg=disabled -Dlcms2:tiff=disabled \
    -Dlibusb:tests=false -Dlibusb:examples=false -Dlibplacebo:demos=false \
    -Dlibplacebo:lcms=enabled -Dlibplacebo:shaderc=enabled -Dlibplacebo:d3d11=enabled \
    -Dvulkan=enabled -Dd3d11=enabled -Dxxhash:inline-all=true -Dxxhash:cli=false \
    -Ddrm=disabled -Dlibarchive=disabled -Drubberband=disabled -Dwayland=disabled \
    -Dx11=disabled -Djavascript=enabled \
    --cross-file /home/x86_64-w64-mingw32.txt build`
    
**Errors:**
Run-time dependency shaderc found: NO (tried pkgconfig, pkgconfig, pkgconfig and system)
Run-time dependency libsixel found: NO (tried pkgconfig and cmake)
Run-time dependency spirv-cross-c-shared found: NO (tried pkgconfig and cmake)

meson.build:1005:28: ERROR: Feature d3d11 cannot be enabled: Either is not a win32 desktop or shaderc nor spirv-cross were found!

meson.build:1285:9: ERROR: Dependency lookup for vulkan with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.

subprojects/libplacebo/src/glsl/meson.build:2:10: ERROR: Dependency lookup for shaderc with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.
subprojects/libplacebo/src/d3d11/meson.build:8:2: ERROR: Dependency lookup for spirv-cross-c-shared with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.

**Cross File:**
[binaries]
c = 'x86_64-w64-mingw32-gcc'
cpp = 'x86_64-w64-mingw32-g++'
ar = 'x86_64-w64-mingw32-ar'
strip = 'x86_64-w64-mingw32-strip'
windres = 'x86_64-w64-mingw32-windres'
dlltool ='x86_64-w64-mingw32-dlltool'
exe_wrapper = '/usr/lib/wine/wine64'

[properties]
needs_exe_wrapper = true

[host_machine]
system = 'windows'
cpu_family = 'x86_64'
cpu = 'x86_64'
endian = 'little'

**Attempts:**

    Added pkg-config but it's not compiling statically and seems to be using the root pkg-config instead of MinGW.

**Request:**
Could someone provide guidance on how to successfully enable shaderc, d3d11, and vulkan features for cross-compiling MPV with the above configuration? Any help or suggestions would be greatly appreciated.

@kasper93 Could you please suggest as i checked you are the one who updated windows cross compile doc 

observation :
difference 
debug result (MPV built by here https://sourceforge.net/projects/mpv-player-windows/ ) - working well 

**Using hardware decoding (d3d11va)**.
 **(+) Video --vid=1 (*) (h264 3840x2160 30.000fps)**       
**VO: [gpu] 2560x1080 d3d11[nv12]**
**[KV: 00:00:00 / 00:00:04 (0%)**

MPV built by me 
 
**Using hardware decoding (d3d11va-copy).**
**â— Video --vid=1  (*) (h264 3840x2160 30 fps)**
**VO: [gpu] 2560x1080 nv12**
**V: 00:00:00 / 00:00:04 (0%)**

thank you

**Comment by Andarwinux** [comment]
meson wrapdb doesn't have shaderc yet, so you'll have to build it yourself or create a wrap for it (you can refer to [win32 ci](https://github.com/mpv-player/mpv/blob/master/ci/build-win32.ps1#L43))

**Reply by Windev2024** [reply]
i have tried using given link  for vulkan and spirv-cross but got below error
Run-time dependency spirv-cross-c-shared found: NO (tried pkgconfig and cmake)
meson.build:1285:9: ERROR: Dependency lookup for vulkan with method 'pkgconfig' failed: Pkg-config for machine host machine not found. Giving up.

below sample wrap file 
[wrap-git]
url = https://github.com/KhronosGroup/Vulkan-Loader
revision = main
depth = 1
clone-recursive = true
Method = cmake


I have integrated mpv into my application, but it is crashing due to a buffer overflow issue. The crash details are:

    Fault Module Name: KERNELBASE.dll
    Exception Code: e24c4a02

I suspect the issue is related to output buffering. Adding output redirection seems to prevent the crash, but this does not align with the behavior of the official mpv build. Additionally, I believe using Direct3D 11 (d3d11) might resolve the issue.

observation :
difference
debug result (MPV built by here https://sourceforge.net/projects/mpv-player-windows/ ) - working well

Using hardware decoding (d3d11va).
(+) Video --vid=1 (*) (h264 3840x2160 30.000fps)
VO: [gpu] 2560x1080 d3d11[nv12]
[KV: 00:00:00 / 00:00:04 (0%)

MPV built by me

Using hardware decoding (d3d11va-copy).
â—� Video --vid=1 (*) (h264 3840x2160 30 fps)
VO: [gpu] 2560x1080 nv12
V: 00:00:00 / 00:00:04 (0%)

**Comment by Windev2024** [comment]
@kasper93 and @Andarwinux 
I am trying to build these projects using CMake by creating Meson wrap files and wrap-git for **Lua 5.1(make is needed - created my own meson build but not overriding dependency error ) , spirv-cross, shaderc, and vulkan**, but I am not able to achieve the desired output. Any suggestions? Below are my steps:
 Added this line in the MPV Meson build file:


   
    **meson**
      subproject('lua-5.1', default_options: ['buildtype=debug'])



Cross file configuration:

    [binaries]
    c = 'i686-w64-mingw32-gcc'
    cpp = 'i686-w64-mingw32-g++'
    ar = 'i686-w64-mingw32-ar'
    strip = 'i686-w64-mingw32-strip'
    windres = 'i686-w64-mingw32-windres'
    dlltool = 'i686-w64-mingw32-dlltool'
    ld = 'i686-w64-mingw32-ld'
    cmake = '/usr/bin/cmake'

    [built-in options]
    buildtype = 'release'

    [host_machine]
    system = 'windows'
    subsystem = 'windows'
    kernel = 'nt'
    cpu_family = 'x86'
    cpu = 'i686'
    endian = 'little'

**Reply by Windev2024** [reply]
> @kasper93 and @Andarwinux I am trying to build these projects using CMake by creating Meson wrap files and wrap-git for **Lua 5.1(make is needed - created my own meson build but not overriding dependency error ) , spirv-cross, shaderc, and vulkan**, but I am not able to achieve the desired output. Any suggestions? Below are my steps: Added this line in the MPV Meson build file:
> 
> ```
> **meson**
>   subproject('lua-5.1', default_options: ['buildtype=debug'])

I have achieved this part Lua 5.1 and Lua 5.2 by creating meson build file 
> ```
> 
> Cross file configuration:
> 
> ```
> [binaries]
> c = 'i686-w64-mingw32-gcc'
> cpp = 'i686-w64-mingw32-g++'
> ar = 'i686-w64-mingw32-ar'
> strip = 'i686-w64-mingw32-strip'
> windres = 'i686-w64-mingw32-windres'
> dlltool = 'i686-w64-mingw32-dlltool'
> ld = 'i686-w64-mingw32-ld'
> cmake = '/usr/bin/cmake'
> 
> [built-in options]
> buildtype = 'release'
> 
> [host_machine]
> system = 'windows'
> subsystem = 'windows'
> kernel = 'nt'
> cpu_family = 'x86'
> cpu = 'i686'
> endian = 'little'
> ```

**Comment by kasper93** [comment]
> meson.build:1005:28: ERROR: Feature d3d11 cannot be enabled: Either is not a win32 desktop or shaderc nor spirv-cross were found!

You need to build shaderc and spirv-cross and add them to your build root. Both the mingw and win32 CI builds do this.

> Added pkg-config but it's not compiling statically and seems to be using the root pkg-config instead of MinGW.

You need to configure pkg-config to use your cross sysroot. Please refer to `PKG_CONFIG_SYSROOT_DIR` and `PKG_CONFIG_LIBDIR`.

> Could someone provide guidance on how to successfully enable shaderc, d3d11, and vulkan features for cross-compiling MPV with the above configuration?

I'm afraid not. We, as the mpv team, are not able to provide individual technical support for cross-compiling software. The compile-windows.md guide already contains extensive details. There are multiple scripts available that automate the tasks you are asking about. You can find these at https://github.com/mpv-player/mpv/blob/master/ci.

All the necessary build steps are standard for this task. There are numerous guides and documentation available online. You can even ask some LLM, it will give you full solution.

I don't mean to sound harsh, but while we have provided a guide on how to start and build mpv, we cannot provide a guide for building every single dependency you might want to use. It is not feasible. All issues you are facing are not related to mpv in any way, you can try to ask in respective dependencies issue trackers, if they are willing to help cross compile. Also they might have already similar guides as mpv.

I encourage others to help, but I personally do not have additional advice on this topic.

> meson wrapdb doesn't have shaderc yet, so you'll have to build it yourself or create a wrap for it (you can refer to [win32 ci](https://github.com/mpv-player/mpv/blob/master/ci/build-win32.ps1#L43))

This is possible, but except for build testing, I would not recommend using these wrappers for production builds. Using the upstream build system is better, as CMake support in Meson is subpar. For guidance on how to do it with pkg-config, please refer to https://github.com/mpv-player/mpv/blob/master/ci/build-mingw64.sh.

Note that if there is upstream wrap for some dependency using subproject is fine, but wrapping CMake is tricky.

> Lua

There is luajit wrap that will work out of the box. `-Dlua=luajit` or you can build Lua manually.

**Reply by Windev2024** [reply]
Thanks 🙏
For lua I have found solution
 referred meson wrap lua4.6 and wrote meson build it's working fine
